Subject: Linux-RT 2.6.25.4-RT
From: http://www.kernel.org/pub/linux/kernel/projects/rt/
Acked-by: Sven-Thorsten Dietrich <sdietrich@suse.de>
From h-shimamoto@ct.jp.nec.com Thu May 15 10:15:00 2008
Date: Mon, 28 Apr 2008 11:17:48 -0700
From: Hiroshi Shimamoto <h-shimamoto@ct.jp.nec.com>
To: Ingo Molnar <mingo@elte.hu>, Steven Rostedt <rostedt@goodmis.org>,
     Thomas Gleixner <tglx@linutronix.de>
Cc: linux-kernel@vger.kernel.org, linux-rt-users@vger.kernel.org
Subject: [PATCH -rt 3/4] x86: nmi_watchdog NMI needed for
    irq_show_regs_callback()

From: Hiroshi Shimamoto <h-shimamoto@ct.jp.nec.com>

The -rt kernel doesn't panic immediately when NMI lockup detected.
Because the kernel waits show_regs on all cpus, but NMI is not come so
frequently.

Signed-off-by: Hiroshi Shimamoto <h-shimamoto@ct.jp.nec.com>
---
 arch/x86/kernel/nmi_32.c |    7 +++++++
 arch/x86/kernel/nmi_64.c |    8 +++++++-
 2 files changed, 14 insertions(+), 1 deletion(-)

Index: linux-2.6.25.4-rt1/arch/x86/kernel/nmi_32.c
===================================================================
--- linux-2.6.25.4-rt1.orig/arch/x86/kernel/nmi_32.c	2008-05-17 08:28:14.000000000 -0400
+++ linux-2.6.25.4-rt1/arch/x86/kernel/nmi_32.c	2008-05-17 08:28:15.000000000 -0400
@@ -426,6 +426,13 @@ __kprobes int nmi_watchdog_tick(struct p
 				if (i == cpu)
 					continue;
 				nmi_show_regs[i] = 1;
+			}
+
+			smp_send_nmi_allbutself();
+
+			for_each_online_cpu(i) {
+				if (i == cpu)
+					continue;
 				while (nmi_show_regs[i] == 1)
 					cpu_relax();
 			}
Index: linux-2.6.25.4-rt1/arch/x86/kernel/nmi_64.c
===================================================================
--- linux-2.6.25.4-rt1.orig/arch/x86/kernel/nmi_64.c	2008-05-17 08:28:14.000000000 -0400
+++ linux-2.6.25.4-rt1/arch/x86/kernel/nmi_64.c	2008-05-17 08:28:15.000000000 -0400
@@ -409,10 +409,16 @@ int __kprobes nmi_watchdog_tick(struct p
 				if (i == cpu)
 					continue;
 				nmi_show_regs[i] = 1;
+			}
+
+			smp_send_nmi_allbutself();
+
+			for_each_online_cpu(i) {
+				if (i == cpu)
+					continue;
 				while (nmi_show_regs[i] == 1)
 					cpu_relax();
 			}
-
 			die_nmi("NMI Watchdog detected LOCKUP on CPU %d\n", regs,
 				panic_on_timeout);
 		}
